### PR TITLE
[Wayland] Handle 0×0 xdg_toplevel_configure in fullscreen

### PR DESCRIPTION
**Pull request by sulix**
tl;dr: Wayland fullscreen is broken under KDE Plasma 5.23.4, and crashes with the message:
```
xdg_surface@28: error -1: invalid window geometry size (0x0)
```

The xdg_shell spec seems to state[1] that ``xdg_toplevel_configure`` events can always provide a 0×0 width/height to signal that the compositor doesn't care. SDL previously assumed the provided width/height was always valid for fullscreen windows, and so applied it as-is.

This broke SDL applications on KDE/KWin 5.23, which now sends 0×0 configure events (and, in 5.23.3, 1×1 events for some reason), breaking all SDL applications in fullscreen[2], when ``SDL_VIDEODRIVER=wayland``

See also the mention of the 1×1 window issue in https://github.com/libsdl-org/SDL/pull/4306#issuecomment-976272935

[1]: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/6
[2]: https://bugs.kde.org/show_bug.cgi?id=444962#c6

### CODE REVIEW

**Comment by flibitijibibo** [comment]
We used to do this but we were eventually instructed to do what is now the current behavior. I'm going to need a signoff from at least one compositor developer before I flip-flip once again on this configuration stuff.

**Comment by flibitijibibo** [comment]
Previous commit, where we only got away with doing this because the event was completely stateless: https://github.com/libsdl-org/SDL/commit/ae7ee2a59c8d41d67c6bc80f25b6e52670dddbbf

**Comment by flibitijibibo** [comment]
One more citation: https://github.com/libsdl-org/SDL/pull/4598#issuecomment-893629121

**Comment by flibitijibibo** [comment]
Found the exact text that we were spec-lawyered into doing this:

> The surface is fullscreen. The window geometry specified in the configure event is a maximum; the client cannot resize beyond it. For a surface to cover the whole fullscreened area, the geometry dimensions must be obeyed by the client. For more details, see xdg_toplevel.set_fullscreen.

We were told to unconditionally treat the numbers as a maximum - and since it's fullscreen, we just take the max values as the size. _Apparently_ this overrides the specification stating 0, 0 allows for flexibility, because when we do that we apparently get crashes on Weston saying our fullscreen window is too big (because woops, the configure doesn't tell us what output we ended up on even though fullscreen has an output as a parameter), because apparently that's a thing that can happen.

**Comment by sulix** [comment]
Huh…

It doesn't make sense to me that this should override the (0,0) "don't care" value: treating (0,0) as a maximum window size seems pretty useless in practice.

I guess there are two questions I have:
1. In what circumstances should a compositor send a (0, 0) configure to a fullscreen window? If the answer is "only when the display is actually 0×0 pixels", then KWin has a bug here.
2. If we get a (0, 0) configure, what should we do? The options seem to be: ignore the whole configure event, respond by setting the size to (0,0), or respond by setting the size to some other size (probably the current size, or the current output size, or something).

With the status quo, there's definitely a bug somewhere between us and KWin. My reading of the spec didn't seem to suggest that fullscreen should override the "don't care" interpretation of (0, 0), but it's definitely possible I'm missing something.

And, for what it's worth, I've just tried this patch on Weston (albeit, just a simple nested session), and it seems to work fine…

**Comment by flibitijibibo** [comment]
1. I want to say when the display is 0,0 but you're right in that it's completely impractical and just makes the expected behavior unclear - in my head I would just ignore it exactly as your patch does, but...
2. The first and third options, which I think are both valid, apparently have issues if the display size isn't what we're allowed to do.

From what I remember, the bug is this:

- Set fullscreen on output A
- Fullscreen is actually B, which is smaller, but we don't have the enter event yet
- We see 0,0 and ignore
- We set geometry for A's size, Weston crashes because it disallows sizes larger than B

But as with most protocol-related bugs I may be misremembering. Like I said, to me your patch makes total sense but I would like to have a server-side developer sign off on it so we can say we made the decision with consultation from both sides of the protocol.

**Comment by slouken** [comment]
We're running into this with the Steam Link app on some configurations. I'm going to go ahead and apply it and we can have a server side developer sign off after the fact.

**Comment by zzag** [comment]
It's a kwin bug, should be fixed in 5.23.5.

**Comment by flibitijibibo** [comment]
Plasma 5.23.5 was released earlier this week:

https://kde.org/announcements/changelogs/plasma/5/5.23.4-5.23.5/

**Comment by dos1** [comment]
I know I'm super late, but just to make this thing clear:

> In what circumstances should a compositor send a (0, 0) configure to a fullscreen window?

In none. If the compositor sends a (0,0) configure with fullscreen state set, it is 100% compositor bug.
